### PR TITLE
60787 anti spam controls

### PR DESCRIPTION
**Pull request by jrjamespdx**
## Summary

This PR further expands upon our initial work of creating a temporary service to send out login credential adoption related emails to eligible veteran users.  The main goal of this PR is to add functionality to prevent us from repeatedly sending emails to veterans within a one (1) week time frame since their last login that triggered an email send.

Put slightly more plainly:  While users may log in multiple times within a one (1) week period, with the work in this PR they should only receive one of these login action triggered Credential Adoption emails once every seven (7) days.

In specific, this PR makes the following changes:
- adds additional methods in the AcceptableVerifiedCredentialAdoptionService to create and query the new database table to record and then later evaluate if a specific user has previously triggered an email send within the last seven (7) days.
- updates/refactors some of the tests
- updates/refactors some of the naming conventions used in the service and tests
- adds a factory for test usage
- adds a new model (with validations) in accordance with the aforementioned new database table

The basic flow of this solution is as follows:
- once called, the AcceptableVerifiedCredentialAdoptionService evaluates if the veteran user has the appropriate credentials to be considered eligible to receive the Credential Adoption email.  This evaluation consists of:
  - checking if the user has both a DSLogon (legacy) and Login.gov (modern) credential
  - checking if the user's most recent login was with the legacy or modern credential 
  - checking if there is a record in the new database table with an email_triggered_at timestamp within the last seven (7) days
- If the user meets the eligibility criteria, we process the message for sending via VA Notify and then capture the appropriate events:
  - two (2) email send action events logged via StatsD
  - one (1) new record created in the CredentialAdoptionEmailRecords table
- If the user does not meet the first two (2) evaluation criteria (possessing both credentials and using the legacy credential most recently), the flow stops and nothing more happens.

_Please note:  The code changes required to call this service via the AfterLoginActions service will be made in a separate PR after this one is fully reviewed, approved, and merged in to main/master._

This work is being done by the Login.gov Adoption Team which reports to John Rahaghi, Product Owner with OCTO on the VA side.

## Related issue(s)
- [ZenHub Ticket 60787](https://app.zenhub.com/workspaces/logingov-adoption-team-632280b31e745000136a61fc/issues/gh/department-of-veterans-affairs/va.gov-team/60787)

## Testing done

Prior Behavior:
The prior behavior for this feature was that if called, it would only determine the first two (2) eligibility criteria and, if those are met, send the email regardless of how many times it may have been triggered and sent before.

Steps to Verify:
Since this functionality is not currently hooked up to the after_login_actions service, there is no user facing way to directly trigger the AcceptableCredentialAdoptionService perform method.  If an engineer were to use a rails console to manually call this service and pass in a user, that user should be evaluated as expected depending on what eligibility criteria they meet/don't meet.  

## What areas of the site does it impact?
Currently it impacts no user facing portions of the site.  On the non-user facing side, it only makes changes to currently unused backend code and adds a new database table which is not in use by or tied to any other feature/service.

## Acceptance criteria

- [x]  I added and updated unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

**Review by holytoastr — APPROVED** [review]
Per discussion, this looks good to move forward! 👍

**Comment by jrjamespdx** [comment]
@rileyanderson @bosawt 

Hey!  Wanted to see if we could get eyes on this PR this morning, if possible.  Appreciate any help/guidance you all can provide :)

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (bosawt) [inline app/services/acceptable_verified_credential_adoption_service.rb:11]: is there a reason to have this as a setting instead of hard coded here? Since it's a constant it will still require a deploy to change this

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (jrjamespdx) [reply app/services/acceptable_verified_credential_adoption_service.rb:95]: If they logged in again using the legacy credential seven (7) days after the first email was triggered, yes they would then trigger the email to be sent again.

The seven (7) day timeline was somewhat arbitrarily set since we only planned to have this feature enabled for one (1) week at most.  

In the [follow-up ticket/PR](https://github.com/department-of-veterans-affairs/vets-api/pull/13304) to this one, we outline the one (1) week experiment timeline.

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (bosawt) [reply app/services/acceptable_verified_credential_adoption_service.rb:11]: Oh I read the PR description, yes, this appears to be the intention.

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (bosawt) [reply app/services/acceptable_verified_credential_adoption_service.rb:95]: Sounds good, just checking that's the intention

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (jrjamespdx) [reply app/services/acceptable_verified_credential_adoption_service.rb:11]: Initially it was because we anticipated having to send out multiple emails over time via this service.  Now that we've pivoted to using GovDelivery for future sends, this isn't necessary anymore.  

I can remove it and use a hardcoded version or we can keep it as is and it'll be removed once we fully transition to the GovDelivery service in August.

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (bosawt) [reply app/services/acceptable_verified_credential_adoption_service.rb:95]: Oh wait, if the feature is planned for 7 days only, would be it simpler just to check and see if the record exists and to add a timeline at some future point?

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (bosawt) [inline app/services/acceptable_verified_credential_adoption_service.rb:94]: Ooh, something I thought of here, I would probably query by ICN instead of email here. It's a safer field to use that's associated with the user

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (bosawt) [inline app/services/acceptable_verified_credential_adoption_service.rb:85]: ICN will always be available if the user is verified. However email may be blank. So there should probably be a check at the start of all this to 'do nothing' if `user.email` is blank

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (jrjamespdx) [reply app/services/acceptable_verified_credential_adoption_service.rb:94]: I went back and forth on that too!  I see what you mean though...the email address could change while the ICN most likely will not.  The edge cases for the email switching during this limited test are likely limited BUT I'm happy to make the change if you feel strongly about it.

Just let me know :)

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (jrjamespdx) [reply app/services/acceptable_verified_credential_adoption_service.rb:85]: Did not know that!  I'll make the change :)

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (bosawt) [reply app/services/acceptable_verified_credential_adoption_service.rb:94]: yeah I'd recommend ICN, you'll have a way better time in general than trying to retrieve a user based off their email

File: app/models/credential_adoption_email_record.rb
Comment (bosawt) [inline app/models/credential_adoption_email_record.rb:4]: (I guess this would be a schema change too) But I'd prob recommend a uniqueness constraint on ICN and email address

Comment (bosawt) [reply app/models/credential_adoption_email_record.rb:4]: Oh wait, unless you intend to have multiple entries based on different templates

Comment (bosawt) [reply app/models/credential_adoption_email_record.rb:4]: In that case, I'd recommend a uniqueness constraint on the intersection of ICN and email template

**Review by bosawt — COMMENTED** [review]
I think I've represented all the feedback I can think about!

One other thing to consider (this would require updating the schema potentially quite a bit so I'll just present it as an idea), but we do have the `UserCredentialEmail` table, which is a mechanism to store the user's email associated with the credential they are currently logged in as. If I were to design this, I probably would have set up the `CredentialAdoptionEmailRecord` table so that it references the `UserCredentialEmail` table instead of directly storing email. It also encrypts their email so there's no worries about bulk PII leak

**Comment by jrjamespdx** [comment]
> 

The main reason we didn't tie the new DB table to anything else is that we knew we would be tearing all this code back out in mid- to late August or early September.  Given that, it felt slightly safer to keep it more isolated so that when that time comes we can pull the data from the table that we need and then drop the entire table without having to worry about other downstream/upstream affects.

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (jrjamespdx) [reply app/services/acceptable_verified_credential_adoption_service.rb]: I removed it and went hardcoded instead, per your suggestion :)

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (jrjamespdx) [reply app/services/acceptable_verified_credential_adoption_service.rb:95]: @bosawt It would have been simpler, yes :)  If this is a blocking change please let me know, otherwise I'll add it to the follow-up PR after this one (which will be going up as soon as this PR is approved and merged in).

**Review by bosawt — APPROVED** [review]
Looks good, a couple of the outstanding comments here are being deferred to the next PR in this work, but this is fine as they are more nitpick level comments than anything else (and the code in this PR is not reachable from prod until the followup PR is merged)

**Review by jeremy6d — APPROVED** [review]
The one comment I had is pretty trivial and not a blocker. I'm beyond grateful for the comprehensive writeup in the PR which made this a lot easier to review.

File: app/services/acceptable_verified_credential_adoption_service.rb
Comment (jrjamespdx) [reply app/services/acceptable_verified_credential_adoption_service.rb:11]: @jeremy6d That was a request from the VA API Engineers code owners, bosawt specifically, which I believe was driven by the short-lived nature of this feature.

**Pull request merged**
